src/action.py

```diff
@@ -0,0 +1,344 @@
+        discourse: A client to the documentation server.
+        draft_mode: If enabled, only log the action that would be taken.
+        delete_pages: Whether to delete pages that are no longer needed.
+
+    Returns:
+        A report on the outcome of executing the action.
+
+    Raises:
+        ActionError: if an action that is not handled is passed to the function.
+    """
+    match action.action:
+        case types_.Action.CREATE:
+            # To help mypy (same for the rest of the asserts), it is ok if the assert does not run
+            assert isinstance(action, types_.CreateAction)  # nosec
+            return _create(action=action, discourse=discourse, draft_mode=draft_mode)
+        case types_.Action.NOOP:
+            assert isinstance(action, types_.NoopAction)  # nosec
+            return _noop(action=action)
+        case types_.Action.UPDATE:
+            assert isinstance(action, types_.UpdateAction)  # nosec
+            return _update(action=action, discourse=discourse, draft_mode=draft_mode)
+        case types_.Action.DELETE:
+            assert isinstance(action, types_.DeleteAction)  # nosec
+            return _delete(
+                action=action,
+                discourse=discourse,
+                draft_mode=draft_mode,
+                delete_pages=delete_pages,
+            )
+        # Edge case that should not be possible
+        case _:  # pragma: no cover
+            raise exceptions.ActionError(
+                f"internal error, no implementation for action, {action=!r}"
+            )
+
+
+def _run_index(
+    action: types_.AnyIndexAction, discourse: Discourse, draft_mode: bool
+) -> types_.ActionReport:
+    """Take the index action against the server.
+
+    Args:
+        action: The actions to take.
+        discourse: A client to the documentation server.
+        draft_mode: If enabled, only log the action that would be taken.
+
+    Returns:
+        A report on the outcome of executing the action.
+
+    Raises:
+        ActionError: if an action that is not handled is passed to the function.
+    """
+    logging.info("draft mode: %s, action: %s", draft_mode, action)
+
+    if draft_mode:
+        return types_.ActionReport(
+            table_row=None,
+            url=DRAFT_NAVLINK_LINK,
+            result=types_.ActionResult.SKIP,
+            reason=DRAFT_MODE_REASON,
+        )
+
+    match action.action:
+        case types_.Action.CREATE:
+            try:
+                # To help mypy (same for the rest of the asserts), it is ok if the assert does not
+                # run
+                assert isinstance(action, types_.CreateIndexAction)  # nosec
+                url = discourse.create_topic(title=action.title, content=action.content)
+                return types_.ActionReport(
+                    table_row=None, url=url, result=types_.ActionResult.SUCCESS, reason=None
+                )
+            except exceptions.DiscourseError as exc:
```

Comment:
The report that is returned differs in the two cases, in one the url at the time of review was set to `None` which has since been corrected to be the `FAIL_NAVLINK_LINK` for consistency. The update case sets the url to the existing url for the index page. Arguably the url could just be set within each case, however for the create case the url only becomes available after the discourse `create_topic` call. The URL could be initialised to the fail case, although to me the current code is easier to understand. Happy for other feedback/ other ways of generating the report

Reply:
Just closing this since I believe I addressed it, please let me know if you have further feedback